### PR TITLE
TEP-0118: Added `Matrix.Include` field in preview mode

### PR DESCRIPTION
**Pull request by EmmaMunley**
# Changes
[TEP-0118](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md) proposed adding include as a new field within matrix

This is the first of several PRs to add the feature to allow specify specific combinations or explicit combinations for Matrix Parameters. 

Note: This PR is to only add the new `Matrix.Include` field and update the documentation in "Preview Mode".  No functionality has been added at this time.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [N/A ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
API Change: Added Matrix.Include in preview mode (Not yet functional)
```

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 1]
```diff
@@ -170,6 +170,21 @@ type Matrix struct {
...
 	Params []Param `json:"params,omitempty"`
+
+	// Include is a list of MatrixInclude
+	// Note this field is in preview mode and not yet supported
+	Include []MatrixInclude `json:"include,omitempty"`
+}
+
+// MatrixInclude allows passing in a specific combinations of Parameters into the Matrix.
+// Note this struct is in preview mode and not yet supported
+type MatrixInclude struct {
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:181]: I thought this kind of struct might already be defined somewhere, but didn't find any 🙃

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 2]
```diff
@@ -170,6 +170,21 @@ type Matrix struct {
 	// The names of the `params` in the `Matrix` must match the names of the `params` in the underlying `Task` that they will be substituting.
 	// +listType=atomic
 	Params []Param `json:"params,omitempty"`
+
+	// Include is a list of MatrixInclude
+	// Note this field is in preview mode and not yet supported
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:175]: thank you @EmmaMunley for this PR, it looks good to me!

The build is failing with:

```
ERROR: New API rule violations found which are not present in hack/ignored-openapi-violations.list. Please fix these violations:
API rule violation: list_type_missing,github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1,Matrix,Include
API rule violation: list_type_missing,github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1,MatrixInclude,Params
```

I think we are missing an annotation here:

```
// +listType=atomic
```

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 3]
```diff
@@ -170,6 +170,21 @@ type Matrix struct {
...
+
+// MatrixInclude allows passing in a specific combinations of Parameters into the Matrix.
+// Note this struct is in preview mode and not yet supported
+type MatrixInclude struct {
+	// Name the specified combination
+	Name string `json:"name,omitempty"`
+
+	// Params takes only `Parameters` of type `"string"`
+	// The names of the `params` must match the names of the `params` in the underlying `Task`
+	Params []Param `json:"params,omitempty"`
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:187]: same annotation missing here:

```
// +listType=atomic
```

File: pkg/apis/pipeline/v1beta1/pipeline_types.go (hunk 2 above)
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1beta1/pipeline_types.go:175]: Thanks for the catch!

**Review by EmmaMunley — COMMENTED** [review]
Thanks Priti! I've updated for both v1 and v1beta1.

File: pkg/apis/pipeline/v1beta1/swagger.json [hunk 4]
```diff
@@ -151,174 +151,7 @@
         }
       }
     },
-    "v1alpha1.PipelineResource": {
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/swagger.json:154]: Why these changes are part of this PR? 🤔 They should not belong here, right?

**Comment by EmmaMunley** [comment]
Having issues with the `hack/update-openapigen.sh` script to regenerate files so am closing this out and opening a new PR here https://github.com/tektoncd/pipeline/pull/6188
